src/backend/renderer/damage/shaper.rs

```diff
@@ -0,0 +1,407 @@
+use std::mem;
+
+use crate::utils::{Physical, Rectangle, Size};
+
+/// The recommended minimum tile side.
+const DEFAULT_MIN_TILE_SIDE: i32 = 16;
+
+/// The maximum ratio of the largest damage rectangle to the current damage bbox.
+const MAX_DAMAGE_TO_DAMAGE_BBOX_RATIO: f32 = 0.9;
+
+/// The gap between tiles damage to merge tiles into one.
+const TILE_DAMAGE_GAP_FRACTION: i32 = 4;
+
+/// State of the damage shaping.
+#[derive(Debug, Default)]
+pub struct DamageShaper<const MIN_TILE_SIDE: i32 = DEFAULT_MIN_TILE_SIDE> {
+    /// Cache for tiles to avoid re-allocation.
+    tiles_cache: Vec<Tile>,
+    /// The damage accumulated during shaping.
+    out_damage: Vec<Rectangle<i32, Physical>>,
+}
+
+impl<const MIN_TILE_SIDE: i32> DamageShaper<MIN_TILE_SIDE> {
+    /// Shape damage rectangles.
+    #[profiling::function]
+    pub fn shape_damage(&mut self, in_damage: &mut Vec<Rectangle<i32, Physical>>) {
+        self.out_damage.clear();
+        self.tiles_cache.clear();
+
+        // Call the implementation without direction.
+        self.shape_damage_impl(in_damage, None, false);
+
+        // The shaped damage is inside of `out_damage`, so swap it with `in_damage` since
+        // it's irrelevant.
+        mem::swap(&mut self.out_damage, in_damage);
+    }
+
+    // A divide and conquer hybrid damage shaping algorithm.
+    //
+    // The key idea is to split rectangles by non-overlapping segments on dominating axis (largest
+    // side of the damage bounding box).
+    //
+    // When damage is overlapping through the entire `in_damage` span we shape damage using the
+    // _tile_ based shaper, where the damage bounding box is split into tiles and damage is being
+    // computed for each tile individually by walking all the current damage rectangles.
+    #[inline(never)]
+    fn shape_damage_impl(
+        &mut self,
+        in_damage: &mut [Rectangle<i32, Physical>],
+        last_direction: Option<DamageSplitAxis>,
+        invert_direction: bool,
+    ) {
+        // Optimize small damage input.
+        if in_damage.is_empty() {
+            return;
+        } else if in_damage.len() == 1 {
+            self.out_damage.push(in_damage[0]);
+            return;
+        }
+
+        // Compute the effective damage bounding box and the maximum damaged area rectangle.
+
+        let (x_min, y_min, x_max, y_max, max_damage_area) =
+            in_damage.iter().fold((i32::MAX, i32::MAX, 0, 0, 0), |acc, rect| {
```

Comment:
Are input rects' locations guaranteed to be > 0? My monitor rn is located at negative coordinates.